rfcs/0000-ret-cont-recursive-nix.md

```diff
@@ -0,0 +1,245 @@
+Derivations that that do so indicate this with some special attribute, say `__recursive`.
+Such derivations must have two outputs, `store` and `drv`.
+`store` would be a local Nix store limited to just drvs and fixed output builds.
+`drv` would contain a symlink to one of the derivations in the store, the root.
+After the build completes, Nix verifies all the drv files and fixed outputs are valid (contents match hashes, etc.) and merges the built store into the ambient store.
+Finally, any uses of the original derivation can be substituted to instead use the symlinked derivation.
+
+To faux-formalize everything in the vein of a small-step semantics:
+```
+immediatelyDependsOn(drv0, drv1)
+immediatelyDependsOn(drv1, drv2)
+-------------------------------------------------- deps-trans
+transitivelyDependsOn(drv0, drv2)
+```
+```
+∀<d : transitivelyDependsOn(drv, -)>
+  d ? __recursive == false
+-------------------------------------------------- build-readiness
+isReadyToBuild(drv)
+```
+```
+drv0 : Drv
+∀<o : outputs(drv0)> build_o : StorePath
+∀<o : outputs(drv0)> build(drv0) = { ${o} = build_o; }
+isReadyToBuild(drv0)
+drv0 ? __recursive == false
+-------------------------------------------------- normal-build
+∀<o : outputs(drv0)> assoc(drv0, o, build_o)
+```
+```
+drv0, drv1 : Drv
+drv1path : RelativePath
+∀<o : outputs(drv0)> build0_o : StorePath
+isReadyToBuild(drv0)
+drv0 ? __recursive == true
+drv0.outputs = { "store" = ...; "drv" = ...; }
+build(drv0) = { succeeded = build0; }
+isTrustlessStore(build0_store)
+drv1 = read(build0_store + drv1path)
+readlink(build0_drv) = build0.store + drv1path
+-------------------------------------------------- immediate-drv-deligation
+reducesTo(drv0, drv1)
+```
+```
+drv0, drv1, drv2 : Drv
+reducesTo(drv1, drv2)
+immediatelyDependsOn(drv0, drv1)
+-------------------------------------------------- transitive-drv-deligation
+reducesTo(drv0, drv0[drv2/drv1])
+```
+```
+drv0, drv1 : Drv
+reducesTo(drv0, drv1)
+∀<o : outputs(drv0)> build0_o : StorePath
+∀<o : outputs(drv0)> assoc(drv0, o, build0_o)
+-------------------------------------------------- delegative-build
+∀<o : outputs(drv0)> assoc(drv1, o, build1_o)
+```
+
+## Design Notes
+
+There's a few things we can call out from the faux-formalization.
+
+ - `isTrustlessStore` is called that because the restricted on the contents—fixed output builds / plain data and drvs—is fully and cheaply verifiabled.
+   This is in contrast to normal builds,
+   where the relationship between the derivation and build can only be verified by redoing the build,
+   and where even then there's no way to know whether to blame the output for being actually malicious, or the derivation for merely being non-deterministic.
+
+ - The substitution of drvs in a downstream derivation reminds me of the substitution of drvs for content hashes with the intensional store.
```

Comment:
Yes, this was one of my comments on that PR. We should figure out how to handle reductions more generally: In the continuation-based recursive nix case, we have `foo.drv → foo'.drv` whereas in the intensional store case we have `(foo.drv)!out → cashash-foo`, but the principle is the same.